Add registry builder for Instrument

Doc94 opened the pull request:
This PR aims to provide more fields to the MusicInstrument currently this class with many deprecated methods for the instrument but not provide any method for the fields in instrument https://minecraft.wiki/w/Data_component_format#instrument like the description or duration.

This depends on https://github.com/PaperMC/Paper/pull/11918 for items with instruments custom

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/registry/data/InstrumentRegistryEntry.java:
```diff
@@ -0,0 +1,90 @@
+
+    Component description();
+
+    Sound sound();
```
This is incorrect. We cannot/should not use the Sound type here. This has to be an Either<SoundEventRegistryEntry, TypedKey<Sound>>. See how I've implemented it in JukeboxSoundRegistryEntry. https://github.com/PaperMC/Paper/pull/11805/files#diff-806c67f8d37d5fadd51ad314002172c70d3eff1ce45f2fe2c7219f27b9ad9b07R30

Doc94 replied:
ohhh i see... first time dealing with this. but then this PR now depends in https://github.com/PaperMC/Paper/pull/11805 for the SoundEventRegistryEntry not?

Lulu13022002 commented on paper-api/src/main/java/io/papermc/paper/registry/data/InstrumentRegistryEntry.java:
```diff
@@ -0,0 +1,89 @@
+public interface InstrumentRegistryEntry {
+
+    @Contract(pure = true)
+    @NonNegative
```
This is the wrong annotation, this should be positive and it can be inlined on the same line, same for range and the builder methods. Even if technically those annotations shouldn't even be used for floats.

lynxplay committed:
Begin 1.21.6-pre1

Initial move of patches to rejected and non rejected source patches to
publish to patch-roulette.

lynxplay committed:
Couple of block entities

NoahvdAa committed:
patch batch

NoahvdAa committed:
work

lynxplay committed:
Entity

Spottedleaf committed:
Add MCUtil#toLocation from Folia

Helper methods to convert NMS world/position to Bukkit Location

Spottedleaf committed:
Replace deprecated Thread#getId usage with Thread#threadId

lynxplay committed:
Commands

ExperienceOrb award methods were consolidated to avoid this large of a
conflict in future versions for what is a single override.
Callers are expected to now pass null to the 3rd overload param as well.

Owen1212055 committed:
/net/minecraft/world/entity/decoration/

Note that a method was removed on Armorstand for a bug fix that makes literally zero sense, probably very old.
Also added and fixed the logic for leashes, as any entity can hold leashes but anything can be leashed.

Also very sad hack used to detect if the value was set.

/net/minecraft/world/level/block/entity/trialspawner/

/net/minecraft/server/network

Owen1212055 committed:
net\minecraft\world\entity\projectile\

Owen1212055 committed:
/net/minecraft/world/entity/monster

Some blocks

LayeredCauldronBlock.java.patch has strange diff-- the comment seems to be a red herring.

net/minecraft/network/protocol/login/

Owen1212055 committed:
ServerPlayer, ServerLevel, ServerEntity
net/minecraft/world/entity/vehicle/

Nothing much to note, reworked the loot table inventory serialization a bit

net/minecraft/world/level/

net/minecraft/world/entity/player

Owen1212055 committed:
Other random classes

net/minecraft/core/dispenser/

Owen1212055 committed:
net/minecraft/world/entity/item

random other server classes

Note: Scoreboards may need handling for the waypoints.

/net/minecraft/world/entity/animal/wolf

Also properly mark the item dropped when shearing as force drops

Owen1212055 committed:
Fix some compile errors

Owen1212055 committed:
Fix dropped diff

Owen1212055 committed:
readd dropped ack diff

lynxplay committed:
Readd dropped diff

Additionally, add some identifying messages to paper comments added by
our more junior developer during the update process.

lynxplay committed:
Some more comments

lynxplay committed:
Last per file patches

Owen1212055 committed:
Fix compile issues:
- Changes:
Turtle now included in the experience orb dropped
- Todo: See what spigot does for CraftEquippableComponent serialization

Owen1212055 committed:
Fix

Owen1212055 committed:
Fixup whatever happened here

lynxplay committed:
TrialSpawner compile fixes

lynxplay committed:
Some initial adoption of ValueInput/Output

lynxplay committed:
Some more compilation issues

lynxplay committed:
Generator

lynxplay committed:
Test fixes

Mostly around the fact that player inventories now properly support
setting and getting the BODY and SADDLE equipment slot. The slots are
exposed to the API via the extra contents array.

lynxplay committed:
Prepare work on feature patches

lynxplay committed:
14 first feature patches

lynxplay committed:
Hacky EAR 2.0 fixes

EAR yields immunity to entities that are not on the ground and are not
"flying entities". What this "flying entity" filter actually entails
however is undocumented. The previous implementation checked for
FlyingMob, which only covered ghasts and phantoms, leaving entities like
the blaze and wither untouched for *some* reason.

To be improved by consulting the elders (cat).

lynxplay committed:
Fix javadocs in DriedGhast

lynxplay committed:
Readd dropped hunk

lynxplay committed:
1.21.6-pre2

lynxplay committed:
Fix compile issues

Lulu13022002 committed:
run generator

Lulu13022002 committed:
some fixes

Lulu13022002 committed:
track block for lava/conduit damage

dropped diff

lynxplay committed:
Moonrise attempt

Notable changes:
 - Mojang now force loads a 1 radius large block around the player,
   including entities.

Do not smite me leaf.

lynxplay committed:
Rest of feature patches

Needs indepth testing of collision improvements on player move packet.

lynxplay committed:
Moonrise compile fixes and changes

Kept in separate patch for easier revert once leaf smites me.

lynxplay committed:
Fix LayeredCauldronBlock event call

Hello masmc05 o/

Spottedleaf committed:
Update Moonrise to 1.21.6-pre2

See https://github.com/Tuinity/Moonrise/commit/b312be292122eef21cd8e617b5015f3e6f020eec

Spottedleaf committed:
Add DataConverter patch

lynxplay committed:
Move over to non discarding reporter

lynxplay committed:
Update ItemType

lynxplay committed:
1.21.6-pre3

lynxplay committed:
Bump server compile memory

jpenilla committed:
Build updates

lynxplay committed:
Add new attributes for 1.21.6

lynxplay committed:
Pass hand to entity unleash event

lynxplay committed:
Resend leash state

lynxplay committed:
Do not send update packets for saddle/body slots

The player inventory now allows setting the saddle and body equipment
slot. The client menu for this however does not contain slots. As such,
updating the player for such slots leads to an IOOB exception.

lynxplay committed:
Second part of diff

Tired.

Spottedleaf committed:
Optimise CraftWorld#getLoadedChunks

We can use the existing full chunk map so that we do not need
to iterate over all ChunkHolders. Additionally, do not use streams
to do the iteration either.

Lulu13022002 committed:
add new UI sound category

lynxplay committed:
Run generators

lynxplay committed:
Remove useless comments and FQN in source

Lulu13022002 committed:
fix some NPEs

Spottedleaf committed:
Update CraftWorld#getForceLoadedChunks to avoid using getChunkAt

Usual methods of adding force loaded chunks sync load them, so
they should be loaded already. This should avoid the more expensive
getChunkAt and more importantly allow this function to properly work
on Folia due to avoiding thread checks.

Lulu13022002 committed:
[ci skip] fix classpath split on Windows for generator CLI

Owen1212055 committed:
Improve command permission lookups

This allows us to append to the vanilla permission check object to avoid having to look into a map.

This also fixes non vanilla commands from technically being able to be skipped through permissions.

Owen1212055 committed:
Prune the api version suffix

This is needed for the test plugin, as the bukkit api does not support these versions in the plugin manifest.

Alternatively we could add another gradle property

Owen1212055 committed:
Remove SERVER_ONLY command flag functionality

This will throw a warning for players on 1.21.6

Owen1212055 committed:
Properly send over SADDLE and BODY equipment items to PlayerInventory

This uses the ClientboundSetPlayerInventoryPacket... in general this logic should be investigated and seen if it can be cleaned up by using this packet.

Owen1212055 committed:
Remove craft flying

Owen1212055 committed:
Removed inlined art api

Inlined art is no longer supported.

Owen1212055 committed:
Add api for shear equippable logic

Owen1212055 committed:
Change ordering on YamlSanitizationTest

DataConverter changes the output of these, but order of the components is generally undefined so just change this for now.

Owen1212055 committed:
Fix V4307 ordering issues

The underlying components seem to still be appended in a different order compared to vanilla but that is fine as the order does not really matter in an ItemStack's component map.

However, 'hidden_components' seem to have defined order that must be respected. We use a hashset and keep the fixing order that is on the vanilla datafixer.

Lulu13022002 committed:
fix sponge absorb resetting block entity data of waterlogged blocks

alternative to https://github.com/PaperMC/Paper/commit/320f25cb04f183c18c16595100897622c88ce073, this commit has the side effect of
not including the waterlogged block into the block list for the later event. Generally this logic is fragile and should be handled by
the BlockStateListPopulator in the future.

Lulu13022002 committed:
support block state mutation of already placed block entity in BlockStateListPopulator

As seen in https://github.com/PaperMC/Paper/issues/12645, the BlockStateListPopulator always destroy
block entity data when the block state change. However this is something that should be supported
so plugin can retrieve block entity data of captured blocks.
Additionally only create snapshots at the end of the capture so we don't need to refresh block entity
data for decorator like the beehive, this is possible since multiple capture
at the same position is not supported and will overwrite the previous data anyway.

Lulu13022002 committed:
fix Phantom#setAnchorLocation(null)

Lulu13022002 committed:
remove support of inlined painting variant in registry

Spottedleaf committed:
Fix infinite loop in RegionFile IO

If an exception is thrown during decompress then the read process
would be started again, which of course would eventually throw in
the decompress process.

Lulu13022002 committed:
attribute modifier display

Lulu13022002 committed:
port some generator fixes

- fix a regression for minecart entity type's class
- stable alphanumeric order for fields and data holder fields for CraftBlockData
- properly register bed and conduit block entity metas

Lulu13022002 committed:
update to pre4

Lulu13022002 committed:
properly override removePassenger for happy ghast

Lulu13022002 committed:
happy ghast is a vehicle

Lulu13022002 committed:
fix indices for horse inventories getItem

If the index is shifted for slots after the saddle slot in the setter
then it should be in the getter to stay consistent.

Machine-Maker committed:
rename RegistryFreezeEvent to RegistryComposeEvent

Lulu13022002 committed:
fix Equippable#shearSound

Lulu13022002 committed:
update to rc1

Lulu13022002 committed:
bump built-in datapack version

Lulu13022002 committed:
tweak BlockStateListPopulator

* catch destroyBlock calls in the populator, this is needed when sponge
absorb water around plants.
Some of them are destroyed in the process and can only live underwater and plugin
should get the final state in that case and not the not waterlogged state.
* handle drying process for waterlogged blocks on the living world at the end,
while the populator will catch any block changes, it will however not trigger
the particle effects done in destroyBlock
* always try to set block entity data even if the underlying block state doesn't
change for snapshot exposed by the events

Lulu13022002 committed:
tweak BlockStateListPopulator pt 2

Lulu13022002 committed:
fix EntityPortalExitEvent handling

Lulu13022002 committed:
do not call BlockGrowEvent on each age grow for cactus

this might be something we do in the future but it should be consistent with the rest too
closes https://github.com/PaperMC/Paper/issues/12662

Lulu13022002 committed:
pass correct context to CraftChatMessage#toJSON

Lulu13022002 committed:
do not teleport player twice on first spawn

Mojang moved the adjustSpawnLocation call from the ServerPlayer constructor
to where we do it normally making it duplicate

Machine-Maker committed:
Add registry builders for SoundEvent and JukeboxSong (#11805)

Machine-Maker committed:
keep original generics on now-deprecated freeze

jpenilla committed:
Disable oldPaper base

Lulu13022002 committed:
Add registry builder for Instrument

Co-authored-by: Pedro <3602279+Doc94@users.noreply.github.com>
